libraries/ts-command-line/src/providers/ScopedCommandLineAction.ts

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { CommandLineAction, ICommandLineActionOptions } from './CommandLineAction';
+import { CommandLineParser, ICommandLineParserOptions } from './CommandLineParser';
+import { CommandLineParserExitError } from './CommandLineParserExitError';
+import type { CommandLineParameter } from '../parameters/BaseClasses';
+import type { CommandLineParameterProvider, ICommandLineParserData } from './CommandLineParameterProvider';
+
+interface IInternalScopedCommandLineParserOptions extends ICommandLineParserOptions {
+  readonly actionOptions: ICommandLineActionOptions;
+  readonly unscopedActionParameters: ReadonlyArray<CommandLineParameter>;
+  readonly onDefineScopedParameters: (commandLineParameterProvider: CommandLineParameterProvider) => void;
+  readonly onExecute: () => Promise<void>;
+}
+
+class InternalScopedCommandLineParser extends CommandLineParser {
+  private _internalOptions: IInternalScopedCommandLineParserOptions;
+
+  public constructor(options: IInternalScopedCommandLineParserOptions) {
+    // We can run the parser directly because we are not going to use it for any other actions,
+    // so construct a special options object to make the "--help" text more useful.
+    const scopingArgs: string[] = [];
+    for (const parameter of options.unscopedActionParameters) {
+      parameter.appendToArgList(scopingArgs);
+    }
+    const unscopedToolName: string = `${options.toolFilename} ${options.actionOptions.actionName}`;
+    const scopedCommandLineParserOptions: ICommandLineParserOptions = {
+      toolFilename: `${unscopedToolName}${scopingArgs.length ? ' ' + scopingArgs.join(' ') : ''} --`,
+      toolDescription: options.actionOptions.documentation,
+      toolEpilog: `For more information on available unscoped parameters, use "${unscopedToolName} --help"`,
+      enableTabCompletionAction: false
+    };
+
+    super(scopedCommandLineParserOptions);
+    this._internalOptions = options;
+    this._internalOptions.onDefineScopedParameters(this);
+  }
+
+  protected onDefineParameters(): void {
+    // No-op. Parameters are manually defined in the constructor.
+  }
+
+  protected async onExecute(): Promise<void> {
+    await super.onExecute();
+    // Redirect action execution to the provided callback
+    await this._internalOptions.onExecute();
```

Comment:
If a consumer's `onExecute` throws, their stacktrace will include `InternalScopedCommandLineParser`. Can we see if there's a way to avoid that?